scripts/shared/e2e.sh

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+
+## Process command line flags ##
+
+source /usr/share/shflags/shflags
+DEFINE_string 'cluster_settings' '' "Settings file to customize cluster deployments"
+DEFINE_string 'focus' '.*' 'Ginkgo focus for the E2E tests'
+FLAGS "$@" || exit $?
+eval set -- "${FLAGS_ARGV}"
+
+focus="${FLAGS_focus}"
+cluster_settings="${FLAGS_cluster_settings}"
+
+set -em -o pipefail
+
+source ${SCRIPTS_DIR}/lib/debug_functions
+source ${SCRIPTS_DIR}/lib/utils
+
+# Always source the shared cluster settings, to set defaults in case something wasn't set in the provided settings
+source "${SCRIPTS_DIR}/lib/cluster_settings"
+[[ -z "${cluster_settings}" ]] || source ${cluster_settings}
+
+### Functions ###
+
+function deploy_env_once() {
+    if with_context cluster3 kubectl wait --for=condition=Ready pods -l app=submariner-engine -n "${SUBM_NS}" --timeout=3s > /dev/null 2>&1; then
+        echo "Submariner already deployed, skipping deployment..."
+        return
+    fi
+
+    make deploy
+}
+
+function generate_context_flags() {
+    for cluster in "${!cluster_subm[@]}"; do
+        if [[ "${cluster_subm[$cluster]}" = "true" ]]; then
+            printf " -dp-context $cluster"
+        fi
+    done
```

Comment:
I don't think this will generally work as expected for all use cases. First `cluster_subm` iteration order is random - submariner, eg,  expects a specific  order: `cluster2 cluster3 cluster1`.  Perhaps we can utilize the ordering of the `clusters` array being introduced by https://github.com/submariner-io/shipyard/pull/113. Second, I don't think we can assume that if `cluster_subm$cluster]` is false then the cluster shouldn't be passed in.

Reply:
Yeah if it makes sense I don't mind passing all the clusters.
However I thought it is expecting only ones that have subm deployed?

As for the order I'm not quite sure why that matters to the e2e tests, shouldn't they work with any order? If we need specific cluster to identify a specific condition (eg gateway failover) then we should probably have an extra flag?

I could change this to receive the clusters order as parameters instead of trying to guess it, in which case the caller would be responsible (and would know it for sure since he's defining the deployment anyhow).

WDYT?

BTW it seems that associative arrays (such as the one used) have consistent order (but not easily predictable)

Reply:
They're passed in a particular order so the code can distinguish them as they may be used for a certain purpose or the code may assume a certain configuration. For submariner, `ClusterC` in the code is used specifically for the add cluster test which assumes no gateway is configured. Lighthouse passes the 3 clusters in numeric order, not sure what assumptions it makes.

I think it's reasonable to use the ordering of the new `clusters` array as the order they will be passed to the test code. So submariner would specify `clusters=(cluster2, cluster3, cluster1)`. Of course then we'll have reconcile positional assumptions in the deploy scripts, eg where it uses `${clusters[1]}` and `${clusters[2]}` for the connectivity test. I assume this is b/c "cluster1" doesn't have submariner running. Actually I'm not clear why we even run the connectivity test on deploy as the E2E tests do that anyway. It seems the scripts should just install/ deploy the clusters and components and let E2E handle the testing. Removing it would certainly simplify the generic scripting.

Reply:
I think the safest is to receive them in the arguments then each project can define it's own order (since they define their own potential cluster settings so they know best what order to send).

Reply:
I agree argument would make sense. Otherwise we need to handle the logic of which one is configured as cluster etc...

e2e_cluster_contexts=(cluster2, cluster3, cluster1)  ? 

something like that